python/ray/autoscaler/_private/spark/node_provider.py

```diff
@@ -0,0 +1,210 @@
+# we can associate launched nodes with their resource reports. IDs increment
+# starting with fffff*00000000 for the head node, fffff*00000001, etc. for workers.
+RAY_ON_SPARK_HEAD_NODE_ID = 0
+RAY_ON_SPARK_HEAD_NODE_TYPE = "ray.head.default"
+
+
+class RayOnSparkNodeProvider(NodeProvider):
+    """A node provider that implements provider for nodes of Ray on spark."""
+
+    def __init__(self, provider_config, cluster_name):
+        NodeProvider.__init__(self, provider_config, cluster_name)
+        self.lock = RLock()
+
+        self._nodes = {
+            RAY_ON_SPARK_HEAD_NODE_ID: {
+                "tags": {
+                    TAG_RAY_NODE_KIND: NODE_KIND_HEAD,
+                    TAG_RAY_USER_NODE_TYPE: RAY_ON_SPARK_HEAD_NODE_TYPE,
+                    TAG_RAY_NODE_NAME: RAY_ON_SPARK_HEAD_NODE_ID,
+                    TAG_RAY_NODE_STATUS: STATUS_UP_TO_DATE,
+                }
+            },
+        }
+        self._next_node_id = 0
+        self.server_url = "https://127.0.0.1:8899"
+
+        self.ray_head_ip = self.provider_config["ray_head_ip"]
+        self.ray_head_port = self.provider_config["ray_head_port"]
+        self.cluster_unique_id = self.provider_config["cluster_unique_id"]
+
+    def get_next_node_id(self):
+        with self.lock:
+            self._next_node_id += 1
+            return self._next_node_id
+
+    def non_terminated_nodes(self, tag_filters):
+        with self.lock:
+            nodes = []
+            for node_id in self._nodes:
+                tags = self.node_tags(node_id)
+                ok = True
+                for k, v in tag_filters.items():
+                    if tags.get(k) != v:
+                        ok = False
+                if ok:
+                    nodes.append(node_id)
+
+            return nodes
+
+    def is_running(self, node_id):
+        with self.lock:
+            return node_id in self._nodes
+
+    def is_terminated(self, node_id):
+        with self.lock:
+            return node_id not in self._nodes
+
+    def node_tags(self, node_id):
+        with self.lock:
+            return self._nodes[node_id]["tags"]
+
+    def _get_ip(self, node_id: str) -> Optional[str]:
+        return node_id
+
+    def external_ip(self, node_id):
+        return self._get_ip(node_id)
+
+    def internal_ip(self, node_id):
+        return self._get_ip(node_id)
+
+    def set_node_tags(self, node_id, tags):
+        assert node_id in self._nodes
+        self._nodes[node_id]["tags"].update(tags)
+
+    def create_node(self, node_config: Dict[str, Any], tags: Dict[str, str], count: int) -> Optional[Dict[str, Any]]:
+        raise AssertionError("This method should not be called.")
+
+    def _gen_spark_job_group_id(self, node_id):
+        return f"ray-cluster-{self.ray_head_port}-{self.cluster_unique_id}-worker-node-{node_id}"
+
+    def create_node_with_resources_and_labels(
+        self, node_config, tags, count, resources, labels
+    ):
```

Comment:
@jjyao I think we should override `create_node_with_resources_and_labels` instead of `create_node` so that we can get the configured resources per worker node from `resources` argument.

Reply:
@jjyao To confirm, does `count` mean the number of nodes to be created ? But seemingly fake multi node cluster neglect this argument

Reply:
> does count mean the number of nodes to be created ?

Yes. IIUC, if you ignore it then you are effectively launching one node at a time so it will be slower.

Reply:
> I think we should override create_node_with_resources_and_labels

Yea, I think it makes sense since we are not using NodeUpdater to start Ray.